String[size()] should return a default constructed CharType

hpvb opened the pull request:
As per the C++ standard 21.3.4.1 for std::string:
Returns: If pos < size(), returns data()[pos]. Otherwise, if pos ==
size(), the const version returns charT(). Otherwise, the behavior is
undefined.

Since the behavior is undefined Godot now does the same thing for const
and non-const versions of operator[].

This fixes #21242 and fixes #22221.

hpvb commented:
PS: for people who watched the stream where this PR was developed, bunnymark performance is in margin of error when not streaming the benchmark :)

reduz commented on core/cowdata.h:
```diff
@@ -52,29 +52,31 @@ class CowData {
 
 private:
 	mutable T *_ptr;
+	T *_defval;
```
I dont think this is a good idea, we should refrain to keep adding stuff here. Also this should be for string, not for cowdata

akien-mga commented:
Added #22221 as fixed by this PR since it seems to be another symptom of the same bug.

akien-mga commented:
Needs a rework to use the new #24385.

akien-mga commented:
Bump. Another occurrence of a `String[size()]` crash: #24585.

hpvb committed:
String[size()] should return a default constructed CharType

As per the C++ standard 21.3.4.1 for std::string:
Returns: If pos < size(), returns data()[pos]. Otherwise, if pos ==
size(), the const version returns charT(). Otherwise, the behavior is
undefined.

Since the behavior is undefined Godot now does the same thing for const
and non-const versions of operator[].

This fixes #21242 and fixes #22221.

akien-mga commented:
Doesn't cherry-pick cleanly on the 3.0 branch, and I'm not sure we should bother at this stage, but I'll let you decide @hpvb.

akien-mga commented:
Opting out from cherry-picking this for 3.0 for now.
